### PR TITLE
Document and fix connection code

### LINKED ISSUE #266: Please support MONGODB_CONNECT setting (true/false)
I had to patch like the following to get the code running on NetBSD, using flask blueprints. 

```
--- connection.py.orig  2016-08-09 16:04:24.000000000 +0200
+++ connection.py   2016-10-09 14:16:12.000000000 +0200
@@ -257,7 +257,7 @@
        # Default to ReadPreference.PRIMARY if no read_preference is supplied
        read_preference = conn_setting.get('read_preference', ReadPreference.PRIMARY)

-        resolved = {}
+        resolved = { 'connect' : False }
        resolved['read_preference'] = read_preference
        resolved['alias'] = alias
        resolved['name'] = db
```

### PR DESCRIPTION
**Pull request by wojcikstefan**
Reverting/ fixing changes from https://github.com/MongoEngine/flask-mongoengine/pull/231. Relevant criticism: https://github.com/MongoEngine/flask-mongoengine/pull/231#issuecomment-255568259.

The main goal of this PR is to fix a buggy connection code introduced in v0.8. It also makes the code cleaner and easier to understand through a better separation of responsibilities and a better documentation.

This is unfortunately dropping support for mongomock, but many things were broken about it to begin with. We should revisit adding it in the future, though personally I don't see a point for it... In the age of Travis/CircleCI/etc., it's so easy to run a real MongoDB in your tests.

Things to test manually:
- [x] Connecting to multiple `mongos` in a sharded cluster
- [x] Connecting to a replica set
- [x] Connecting to a server that requires authentication
- [x] All of the above via a MongoDB URI

Fixes #266 #283 #282 #259 #258 #254 #250 #247 #267

### CODE REVIEW

**Comment by wojcikstefan** [comment]
@lafrech @thedrow lmk what you think + anything you'd add to "Things to test manually"?

**Comment by lafrech** [comment]
LGTM. I did a quick "review". Looked fine. At this point, we can't break it more than it already is, and the best way to get feedback could be to merge into master, maybe even release as 0.8.1, and ping reporters of all related bugs.

> This is unfortunately dropping support for mongomock, but many things were broken about it to begin with. We should revisit adding it in the future, though personally I don't see a point for it... In the age of Travis/CircleCI/etc., it's so easy to run a real MongoDB in your tests.

Totally agree. Why not mongomock if it comes at no cost, but not a priority.

Great job. Thanks @wojcikstefan.

**Comment by wojcikstefan** [comment]
Thanks @lafrech !

> Why not mongomock if it comes at no cost, but not a priority.

IMHO *every* line of code comes at a maintenance cost and if it can be avoided without much of a loss, then it should. Mongomock support is even more baffling, because making one's tests rely on a mocked version of a real database means that your project will always depend on that package staying up-to-date as the real database evolves (which can either mean you'll have to wait longer w/ a mongo upgrade or, worst case, your test will pass and your production environment will break).

**Comment by lafrech** [comment]
Sure. I don't use it myself. I just run a MongoDB instance on my dev machine, and cleanup the DB in fixtures before each test.

[mongomock says](https://github.com/mongomock/mongomock) that this 

> is obviously the best approach here, since we are testing against a real mongodb instance. However, a mongodb instance needs to be set up for this, and cleaned before/after the test. You might want to run your tests in continuous integration servers, on your laptop, or other bizarre platforms - which makes the mongodb requirement a liability.

I guess @losintikfos finds an interest in this. But I agree it sounds like very specific use cases.

**Comment by wojcikstefan** [comment]
I'd love to hear more from people who find an interest in this (including @losintikfos). The fact that MongoDB needs to be cleaned before/after each test is pretty irrelevant given how easy unit testing frameworks make it.

> (...) You might want to run your tests in continuous integration servers, on your laptop, or other bizarre platforms (...)

First, we shouldn't optimize for "bizarre platforms" - there's too much of a cost with connection management to justify some obscure use cases. Second, modern CI servers all support MongoDB:
* Jenkins: https://wiki.jenkins-ci.org/display/JENKINS/MongoDB+Plugin
* Travis: https://docs.travis-ci.com/user/database-setup/#MongoDB
* Circle: https://circleci.com/docs/configuration/#services

Anyway, we should probably take that discussion away from this PR and into a dedicated issue. I'll test the remaining scenarios from this PRs description and then merge/tweak the code.

File: tests/test_connection.py [hunk 1]
```diff
@@ -1,58 +1,19 @@
...
+from mongoengine.context_managers import switch_db
 import pymongo
 from pymongo.errors import InvalidURI
+from pymongo.read_preferences import ReadPreference
 
-from flask_mongoengine import InvalidSettingsError, MongoEngine
-from tests import FlaskMongoEngineTestCase
+from flask_mongoengine import MongoEngine
 
+from .tests import FlaskMongoEngineTestCase
```
Comment (lafrech) [inline tests/test_connection.py]: Looks like this breaks the tests.

https://travis-ci.org/MongoEngine/flask-mongoengine/jobs/178375073#L529

Comment (wojcikstefan) [inline tests/test_connection.py]: Yup, fixed in the latest commit

**Comment by wojcikstefan** [comment]
`test_host_as_list` won't pass until a new mongoengine version is released (due to https://github.com/MongoEngine/mongoengine/pull/1389)

**Comment by wojcikstefan** [comment]
Other than the test issue, I believe this is ready to go. Wanna give it one last glance before merging & releasing @lafrech ?

**Comment by allanlei** [comment]
Instead of deconstructing a connection string and passing the parts to `mongoengine.connect()`, can `flask-mongoengine` *construct* a connection string from the parts and pass to `mongoengine.connect()`?

I find it quite difficult to follow which library is parsing what.  `pymongo` already does parsing of the `host` connection string, wouldn't it be a good idea to delegate the parsing to it?

It would also have the added benefit of seeing the connection string and just trying it out with the mongo CLI.

My original issue #851 which was simply resolved by only using connection string URIs.

**Comment by lafrech** [comment]
>test_host_as_list won't pass until a new mongoengine version is released (due to MongoEngine/mongoengine#1389)

@wojcikstefan, I relaunched the tests with new MongoEngine version out and they passed (except for pypy3, of course).

**Comment by lafrech** [comment]
Thanks @allanlei for your inputs. It sounds like an important rework, so considering flask-mongoengine has been broken for a while now, and this branch fixes numerous issues, I think we should release. It can always be reworked afterwards according to your proposal (and your help on this would be welcome but before coding anything, please open an issue to discuss the implementation, as it would be frustrating to get your code rejected for a design disagreement).

@wojcikstefan, I think we can merge/release. Do you want to do the manual tests before we proceed?

**Comment by wojcikstefan** [comment]
Thank you @allanlei, that's definitely worth looking into. I'm also not a fan of how much abstraction and logic both Flask-MongoEngine AND MongoEngine add on top of PyMongo, but I don't have a clear vision of the fix in my mind yet. We should discuss further in a separate issue (can you open one?).

Agreed @lafrech, this is ready to merge. I performed all the tests I mentioned in the description of this PR already. Can you merge and release? I'll be AFK for the next hour or so.

**Comment by lafrech** [comment]
Merged. I'm in the middle of something, so I don't release right now. Need to double-check changelog and bump version. I don't want to do fast and miss something.

**Pull request merged**